src/compiler/checker.ts

```diff
@@ -16651,8 +16651,18 @@ namespace ts {
             }
         }
 
-        function getTypeOfPropertyOfContextualType(type: Type, name: __String) {
+        function getTypeOfPropertyOfContextualType(type: Type, name: __String, node?: Node) {
             return mapType(type, t => {
+                if (isGenericMappedType(t)) {
+                    const contextualMapper = node ? cloneTypeMapper(getContextualMapper(node)) : identityMapper;

```

Comment:
So this is spooky and likely unsafe. The contextual mapper isn't used during contextual typing because invoking it fixes type parameters because using it implies that you've accepted the inferred type. Cloning avoids the fixing, but not the implication. That implication is important - it's why we fix parameter types to the node when we contextually check them for the first time. Speculatively projecting the result of invoking the mapper without fixing the inference (as this would do) can cause a parameter type to be set _before_ an associated type parameter has been fixed. If that occurs and the inference is tossed (eg, because nothing fixes the inference), then we have a parameter type annotated on the tree informed by a discarded context, and no way to also discard that parameter type (as would be the only safe thing to do).

Reply:
You'd be able to make a test which should demonstrate this with a set of overloads on a function, where checking the first of which triggers this code path where the second should end up as the accepted overload by with a different contextual parameter type applied to the contextually typed function within the mapped type-contextualized parameter.

Reply:
I'm not sure I understand your original comment correctly; you of course know much more about this stuff than I do.  But isn't contextual typing already pretty much best-effort in the presence of multiple overloads?  The TypeScript project seems to be pragmatic in the way it evaluates a lot of things; can you come up with a reasonable scenario in which you get a worse result with this PR than without it?

Reply:
To modify your example:
```ts
declare function f<T extends object>(data: T, handlers: { [P in keyof T]: (value: T[P], prop: P) => string; }): void;
declare function f<T extends object>(data: T, handlers: { [P in keyof T]: (value: object, prop: P) => number; }): void;

f({ data: 0 }, { data(value, key) { return 42; } });
```

Reply:
Is that any different from the following problem that already exists?
```ts
declare function f<T>(data: T, handler: (value: T) => string): void;
declare function f<T>(data: T, handler: (value: object) => number): void;

f(0, (value) => 42);
```

Reply:
Hmmm, you're probably right. I'm just 90% sure I've gotten flak before for using attempting to use a cloned contextual mapper during contextual typing before; but I can;t find whatever PR that conversation happened on. 🤷‍♂️

Reply:
I decided to change the logic for a generic mapped type `{[P in K]: X}` in `getTypeOfPropertyOfContextualType` to mirror the existing (unsound) logic in `resolveMappedTypeMembers` that looks at the base constraint of `K` without instantiating it with the contextual mapper, so we no longer have this controversy.